worlds/jakanddaxter/__init__.py

```diff
@@ -0,0 +1,265 @@
+    can_trade: Callable[[CollectionState, int, Optional[int]], bool]
+    orb_bundle_size: int = 0
+    orb_bundle_item_name: str = ""
+
+    def generate_early(self) -> None:
+        # For the fairness of other players in a multiworld game, enforce some friendly limitations on our options,
+        # so we don't cause chaos during seed generation. These friendly limits should **guarantee** a successful gen.
+        if self.multiworld.players > 1:
+            from .Rules import enforce_multiplayer_limits
+            enforce_multiplayer_limits(self.options)
+
+        # Verify that we didn't overload the trade amounts with more orbs than exist in the world.
+        # This is easy to do by accident even in a single-player world.
+        from .Rules import verify_orb_trade_amounts
+        verify_orb_trade_amounts(self.options)
+
+        # Cache the orb bundle size and item name for quicker reference.
+        if self.options.enable_orbsanity == EnableOrbsanity.option_per_level:
+            self.orb_bundle_size = self.options.level_orbsanity_bundle_size.value
+            self.orb_bundle_item_name = orb_item_table[self.orb_bundle_size]
+        elif self.options.enable_orbsanity == EnableOrbsanity.option_global:
+            self.orb_bundle_size = self.options.global_orbsanity_bundle_size.value
+            self.orb_bundle_item_name = orb_item_table[self.orb_bundle_size]
+
+        # Options drive which trade rules to use, so they need to be setup before we create_regions.
+        from .Rules import set_orb_trade_rule
+        set_orb_trade_rule(self)
+
+    # This will also set Locations, Location access rules, Region access rules, etc.
+    def create_regions(self) -> None:
+        from .Regions import create_regions
+        create_regions(self)
+
+        # from Utils import visualize_regions
+        # visualize_regions(self.multiworld.get_region("Menu", self.player), "jakanddaxter.puml")
+
+    # Helper function to reuse some nasty if/else trees.
+    def item_type_helper(self, item) -> Tuple[int, ItemClassification]:
+        # Make 101 Power Cells.
+        if item in range(jak1_id, jak1_id + Scouts.fly_offset):
+            classification = ItemClassification.progression_skip_balancing
+            count = 101
+
+        # Make 7 Scout Flies per level.
+        elif item in range(jak1_id + Scouts.fly_offset, jak1_id + Specials.special_offset):
+            classification = ItemClassification.progression_skip_balancing
+            count = 7
+
+        # Make only 1 of each Special Item.
+        elif item in range(jak1_id + Specials.special_offset, jak1_id + Caches.orb_cache_offset):
+            classification = ItemClassification.progression
+            count = 1
+
+        # Make only 1 of each Move Item.
+        elif item in range(jak1_id + Caches.orb_cache_offset, jak1_id + Orbs.orb_offset):
+            classification = ItemClassification.progression
+            count = 1
```

Comment:
```suggestion
        elif item in range(jak1_id + Caches.orb_cache_offset, jak1_id + Orbs.orb_offset):
            classification = ItemClassification.progression | ItemClassification.useful
            count = 1
```
This one is a pure preference thing. I'm trying to push for more games to multi-classify some items.
Items can currently have multiple classifications (useful + progression, useful + trap, progression + trap). Obviously, most of these don't have much of a point. But progression + useful could be (and, imo, should be) used to mark something as "the thing the player actually wants", to separate it from stuff like Precursor Orbs or Power Cells.

Basically, clients can already decide to display checks in-game based on classification, and I'd like it to be possible for a game to display locations that have stuff like Moves in a different way than stuff like Orbs.

So, best way to get that going is if any games actually have multiple classifications. Currently, I believe only Mega Man 2 has them, and Tunic has a PR up that changes to have them.

Reply:
Looking at this again, why do orb caches map to moves?

Reply:
I was just thinking about the case where all trade amounts are 0, which would make orbs entirely useless lol

```py
        # Make N Precursor Orb bundles, where N is 2000 / bundle size.
        elif item in range(jak1_id + Orbs.orb_offset, jak1_max):
            if self.total_trade_orbs == 0:
                classification = ItemClassification.filler  # If you don't need orbs to do trades, they are useless.
            else:
                classification = ItemClassification.progression_skip_balancing
            count = 2000 // self.orb_bundle_size if self.orb_bundle_size > 0 else 0  # Don't divide by zero!
```